sparse/coo.py

```diff
@@ -2149,101 +2132,153 @@ def _elemwise_binary(func, self, other, *args, **kwargs):
 
-    self_data = self[self_coords]
+    reduced_coords = [_get_reduced_coords(arg.coords,
+                                          combined_params[-arg.ndim:])
+                      for arg in args]
+    reduced_shape = _get_reduced_shape(args[0].shape,
+                                       combined_params[-args[0].ndim:])
 
-    func_data = func(self_data, other.data, *args, **kwargs)
-    mask = func_data != 0
-    func_data = func_data[mask]
-    func_coords = other.coords[:, mask]
+    reduced_linear = [_linear_loc(coord, reduced_shape)
+                      for coord in reduced_coords]
+    sorted_indices = [np.argsort(lin) for lin in reduced_linear]
 
-    return COO(func_coords, func_data, shape=result_shape,
-               has_duplicates=other.has_duplicates,
-               sorted=other.sorted)
+    reduced_linear = [lin[idx] for lin, idx in zip(reduced_linear, sorted_indices)]
 
+    coords = [arg.coords[:, idx] for arg, idx in zip(args, sorted_indices)]
+    data = [arg.data[idx] for arg, idx in zip(args, sorted_indices)]
 
-def _reverse_self_other(func):
-    def wrapper(*args, **kwargs):
-        return func(args[1], args[0], *args[2:], **kwargs)
+    # Find matches between self.coords and other.coords
+    matched, matched_indices = _nary_match(*reduced_linear)
 
-    return wrapper
+    matched_coords = [c[:, idx] for c, idx in zip(coords, matched_indices)]
 
+    # Add the matched part.
+    matched_coords = _get_matching_coords(matched_coords, broadcast_params, matched_shape)
+    matched_data = [d[idx] for d, idx in zip(data, matched_indices)]
+
+    matched_indices = [sidx[midx] for sidx, midx in zip(sorted_indices, matched_indices)]
+
+    matched_arrays = [COO(matched_coords, d, shape=matched_shape) for d in matched_data]
 
-def _get_unmatched_coords_data(coords, data, shape, result_shape, matched_idx,
-                               matched_coords):
+    return matched_indices, matched_arrays
+
+
+def _unmatch_coo(func, args, mask, **kwargs):
     """
-    Get the unmatched coordinates and data - both those that are unmatched with
-    any point of the other data as well as those which are added because of
-    broadcasting.
+    Matches the coordinates for any number of input :obj:`COO` arrays.
+
+    First computes the matches, then filters out the non-matches.
 
     Parameters
     ----------
-    coords : np.ndarray
-        The coordinates to get the unmatched coordinates from.
-    data : np.ndarray
-        The data corresponding to these coordinates.
-    shape : tuple[int]
-        The shape corresponding to these coordinates.
-    result_shape : tuple[int]
-        The result broadcasting shape.
-    matched_idx : np.ndarray
-        The indices into the coords array where it matches with the other array.
-    matched_coords : np.ndarray
-        The overall coordinates that match from both arrays.
+    args : tuple[COO]
+        The input :obj:`COO` arrays.
```

Comment:
add in `func`, `mask` and `**kwargs` to the docstring?

Reply:
Done!